### PR TITLE
Pointer Constraints and Relative Pointer

### PR DESCRIPTION
**Pull request by Bonicgamer**
This time I did not copy Sway's code, but instead I read Sway's code to get an idea of what I should do or see how Sway solves a problem, closed Sway's code, and then wrote my own code.

Pointer constraints can be tested with the wlroots pointer constraints example.
Relative pointer can be tested with Minetest.

### CODE REVIEW

File: river/Seat.zig [hunk 1]
```diff
@@ -164,9 +165,21 @@ pub fn focus(self: *Self, _target: ?*View) void {
             self.focus_stack.push(new_focus_node);
         }
 
+        if (self.input_manager.pointer_constraints.constraintForSurface(view.surface.?, self.wlr_seat)) |constraint| {
+            @intToPtr(*PointerConstraint, constraint.data).setAsActive();
+        } else if (self.cursor.constraint != null) {
+            PointerConstraint.warpToHint(&self.cursor);
+            self.cursor.constraint.?.sendDeactivated();
+            self.cursor.constraint = null;
+        }
```
Comment (ifreund) [inline river/Seat.zig:174]: This seems strange, why would we want to deactivate the constraint when a surface gains focus?

File: river/Seat.zig [hunk 2]
```diff
@@ -164,9 +165,21 @@ pub fn focus(self: *Self, _target: ?*View) void {
...
         // Focus the target view
         self.setFocusRaw(.{ .view = view });
     } else {
+        if (self.cursor.constraint != null) {
+            PointerConstraint.warpToHint(&self.cursor);
+            self.cursor.constraint.?.sendDeactivated();
+            self.cursor.constraint = null;
```
Comment (ifreund) [inline river/Seat.zig:181]:
```suggestion
        if (self.cursor.constraint) |constraint| {
            PointerConstraint.warpToHint(&self.cursor);
            constraint.sendDeactivated();
            self.cursor.constraint = null;
```

**Review by ifreund — CHANGES_REQUESTED** [review]
Thanks for working on this, seems to work fine in the clients I tested, just some code cleanup to do before this is ready to merge.

File: AUTHORS [hunk 3]
```diff
@@ -3,3 +3,4 @@ The following developers have contributed code to river:
  Leon Henrik Plickat
  Marten Ringwelski
  Rishabh Das
+ Bonicgamer
```
[the remark this answers is not in the document]
Comment (Bonicgamer) [reply AUTHORS:6]: I try not to use my real name for anything.

Also, if you want I could switch it to just "Bonic" if you think that "gamer" is a little too tacky.

[the remark this answers is not in the document]
Comment (ifreund) [reply AUTHORS:6]: Like I said, Bionicgamer is fine if that's what you prefer.

File: river/Seat.zig (hunk 1 above)
Comment (Bonicgamer) [reply river/Seat.zig:174]: I don't think we want the cursor to be constrained to a surface that we are not focusing.
i.e. when switching focus from a game to a web browser.

File: river/PointerConstraint.zig [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        _ = self.constraint.region.intersect(&self.constraint.surface.input_region, &self.constraint.current.region);
+    } else {
+        _ = self.constraint.region.copy(&self.constraint.surface.input_region);
+    }
+    self.checkRegion();
+
+    self.constraint.sendActivated();
+}
+
+fn checkRegion(self: *Self) void {
```
[the remark this answers is not in the document]
Comment (Bonicgamer) [reply river/PointerConstraint.zig:80]: I guess it does.
I'm not good at naming things, so if you think its better then I'd be happy to change it.

**Review by ifreund — APPROVED** [review]
LGTM thanks!

**Comment by FoundOne** [comment]
I'm sorry that I'm late, but I still have an issue with OpenArena. It does not move the mouse at all (it was not like that before). I guess it's some legacy weirdness. It works great in Minetest right now, which is great.

**Pull request merged**